apis/v1/secretproviderclass_types.go

```diff
@@ -46,13 +46,35 @@ type SecretObject struct {
 	Data        []*SecretObjectData `json:"data,omitempty"`
 }
 
+type SyncOptions struct {
+	// type of K8s secret object
+	Type string `json:"type,omitempty"`
+	// the format of the secret (plaintext|json|yaml)
+	Format string `json:"format,omitempty"`
+	// the nested object to target for syncing
+	JsonPath string   `json:"jsonPath,omitempty"`
+	Secrets  []Secret `json:"secrets,omitempty"`
+}
+
+type Secret struct {
+	// type of K8s secret object
+	Type string `json:"type,omitempty"`
+	// name of the K8s secret object
+	SecretName string `json:"secretName,omitempty"`
+	// the format of the secret (plaintext|json)
+	Format string `json:"format,omitempty"`
+	// the nested object to target for syncing
+	JsonPath string `json:"jsonPath,omitempty"`
+}
+
 // SecretProviderClassSpec defines the desired state of SecretProviderClass
 type SecretProviderClassSpec struct {
 	// Configuration for provider name
 	Provider Provider `json:"provider,omitempty"`
 	// Configuration for specific provider
 	Parameters    map[string]string `json:"parameters,omitempty"`
 	SecretObjects []*SecretObject   `json:"secretObjects,omitempty"`
+	SyncOptions   SyncOptions       `json:"syncOptions,omitempty"`
```

Comment:
My major concern here is that this PR adds this great templating/translation feature but _only_ in the context of secret _syncing_.

I would like to see this totally separate from secret syncing as this provides useful transforms for our main usecase of writing secrets to files.

Re-writing this as something like `TransformOptions` then performing the transform between the RPC response from the provider but _before_ the data is written to with the `AtomicWriter` will make this feature useful in many more usecases.

Reply:
Between [`client.Mount()`](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/ac8ae7fc6edf8720c15016bc237a9914db99bbf4/pkg/secrets-store/provider_client.go#L229) RPC to the provider and the [`fileutil.WritePayloads`](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/ac8ae7fc6edf8720c15016bc237a9914db99bbf4/pkg/secrets-store/provider_client.go#L261)

Is I think the idea spot to inject the concept of a "transform" where 1 file can map to multiple files based on json path or other future format extractors.

Reply:
I don't think I am understanding the changes you are looking for @tam7t. My understanding of the issue that you've linked (#948) is that the user wants the ability to hardcode secret k/v pairs directly in `secretObjects`. The functionality I am adding with this feature is meant for targeting a set of secrets when the mounted content is a `json` object with multiple k/v pairs. Can you provide an example to demonstrate the idea behind `the concept of a "transform" where 1 file can map to multiple files based on json path or other future format extractors`?